### PR TITLE
Fix weak object relation check

### LINKED ISSUE #26756: Performance issue checking IndexedAccess rest parameter that resolves to large union of tuples
**TypeScript Version:**  3.1.0-dev.20180829, master

**Search Terms:** rest, parameters, union, tuple, IndexedAccess, 

I realize this may be firmly in the camp of "doctor, it hurts when I do this", but thought I'd still file :)

I ran into this when using a union of tuples in place of overloading, not realizing it wasn't really supposed to be working until #26676. Since the tuples came from the values of an interface, I believe the indirect indexed access was allowing it to still pass since it would fall back to structural checking against `any[]`.

With #26676 landed it's still an issue, though.

**Code**
This was from an attempt to type the Chrome debugger rpc protocol using tuples to represent that some commands need no parameters, some need one, and some can optionally have one. I'll link to the type files since they're large, but I can provide them elsewhere if needed: [`ChromeDevTools/devtools-protocol/pull/113/protocol-mapping.d.ts`](https://github.com/ChromeDevTools/devtools-protocol/blob/68f219dd201c2eb571d31d89a95e8c60c05bd152/types/protocol-mapping.d.ts#L459-L3267), and the types referenced in that file are found in [`protocol.d.ts`](https://github.com/ChromeDevTools/devtools-protocol/blob/68f219dd201c2eb571d31d89a95e8c60c05bd152/types/protocol.d.ts). All the types end up relatively shallow and without cycles.

Poor performance can be triggered with

```ts
import ProtocolMapping  from './protocol-mapping';
type Commands = ProtocolMapping.Commands;

function sendCommand<C extends keyof Commands>(method: C, ...params: Commands[C]['paramsType']) {}
```

This takes ~6 seconds to check with `tsc --strict` on my machine, with about 3.5 seconds attributable to `checkParameter()` of `...params`. More crucially it adds a similar amount of time to tsserver responses, and the work doesn't seem to be cached, so it re-runs on every intellisense completion. 

On the other hand, the not quite equivalent

```ts
function sendCommand<C extends keyof Commands>(method: C, ...params: Commands[keyof Commands]['paramsType']) {}
```

takes about 2 seconds (with about 300 ms of that attributable to `checkParameter` of `...params`).

The difference appears to come down to when the checker realizes the type is a union of tuples. The fast version does appear to end up checking each tuple for structural similarity to `any[]`, but it does so for each tuple in the union individually, which is relatively fast.

The slow version seems to instead get deep enough that it ends up taking the union of each property across the tuples, then compares each of those against the properties of `any[]`. Since in this example basically each tuple ends up with a unique typeArgument, I assume the structural comparison against these unions for each property is what's taking the bulk of this time. The profile is mostly stuck in `getPropertiesOfUnionOrIntersectionType` down to `createUnionOrIntersectionProperty` and then splitting up into many nested `isRelatedTo` calls.

Hopefully the slow case could be made to function more like the fast case, but it also seems like a shortcut is possible since the tuples are never used for anything but rest parameters (so memoized type info is never used again) and as soon as `getApparentType()` is called with `Commands[C]['paramsType']`, the type then becomes a union of objects with `ObjectFlags.Tuple`, which could be used for an even quicker test.

To check, I added a super hacky `isUnionOfTuples` method and the time spent dropped to essentially nothing. I might be missing the corner cases here (or that my case is a special case of some tricky general issue), but if rest parameters for overloading becomes popular, it seems like a fast path for when the apparent type of `source` is a union of tuples and `target === anyArrayType` could be generally useful without much cost.

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR optimizes two code paths:

* Weak object checks now exclude generic types and therefore never fetch constraints. Previously, for a type variable constrained by a union type we'd end up computing the property set of the union type which can be an expensive operation. Also, we'd do this *before* checking whether the target is actually a weak type (a relatively cheap check). Now we do it the other way around.
* Type relationship checks with a tuple on the source side and an array on the target side are now specialized to just check the type of the index signature. This improves performance and error elaborations.

Fixes #26756.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this

File: tests/baselines/reference/tupleTypes.errors.txt [hunk 1]
```diff
@@ -9,15 +9,11 @@ tests/cases/compiler/tupleTypes.ts(18,1): error TS2322: Type '[number, string, n
...
+  Type 'string | number' is not assignable to type 'number'.
+    Type 'string' is not assignable to type 'number'.
 tests/cases/compiler/tupleTypes.ts(49,1): error TS2322: Type '[number, {}]' is not assignable to type 'number[]'.
-  Types of property 'pop' are incompatible.
-    Type '() => number | {}' is not assignable to type '() => number'.
-      Type 'number | {}' is not assignable to type 'number'.
-        Type '{}' is not assignable to type 'number'.
+  Type 'number | {}' is not assignable to type 'number'.
+    Type '{}' is not assignable to type 'number'.
 tests/cases/compiler/tupleTypes.ts(50,1): error TS2322: Type '[number, number]' is not assignable to type '[number, string]'.
```
Comment (sandersn) [inline tests/baselines/reference/tupleTypes.errors.txt]: I looked for explicit tuple-tuple relations in isRelatedTo and couldn't find it. Can you point me to it?

Comment (ahejlsberg) [reply tests/baselines/reference/tupleTypes.errors.txt]: Right here https://github.com/Microsoft/TypeScript/pull/26790/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR11832

**Comment by mihailik** [comment]
> Type relationship checks with a tuple on the source side and an array on the target side are now specialized to just check the type of the index signature. 

What if the array side is **more** specific than just array?

```ts
interface TupleOfMillion extends Array<number> {
    0: 1000000;
    length: 1;
}

var tmillion: TupleOfMillion, tten: [10];

tmillion = tten; // Type '10' is not assignable to type '1000000'.
tten = tmillion; // Type '1000000' is not assignable to type '10'.
```

**Comment by ahejlsberg** [comment]
@mihailik The optimization only applies to checks between tuples and `Array<T>` or `ReadonlyArray<T>` instantiations. All other types are related structurally as usual.
